src/TrancheWrapper.sol

```diff
@@ -0,0 +1,236 @@
+        asset.safeApprove(LENDING_POOL, assets);
+
+        shares = ERC4626(TRANCHE).deposit(assets, address(this));
+
+        _mint(receiver, shares);
+
+        emit Deposit(msg.sender, receiver, assets, shares);
+    }
+
+    /**
+     * @notice Compliant with the standard ERC-4626 mint implementation.
+     * @param shares The amount of shares minted.
+     * @param receiver The address that receives the minted shares.
+     * @return assets The amount of assets of the underlying ERC20 token being deposited.
+     * @dev This contract does not directly transfer the underlying assets from the sender to the receiver.
+     * Instead it calls the deposit of the Lending Pool which calls the transferFrom of the underlying assets.
+     * Hence the sender should not give this contract an allowance to transfer the underlying asset but the Lending Pool instead.
+     */
+    function mint(uint256 shares, address receiver) public override returns (uint256 assets) {
+        assets = ITranche(TRANCHE).previewMintAndSync(shares);
+
+        asset.safeTransferFrom(msg.sender, address(this), assets);
+
+        asset.safeApprove(LENDING_POOL, assets);
+
+        ERC4626(TRANCHE).deposit(assets, address(this));
+
+        _mint(receiver, shares);
+
+        emit Deposit(msg.sender, receiver, assets, shares);
+    }
+
+    /**
+     * @notice Compliant with the standard ERC-4626 withdraw implementation.
+     * @param assets The amount of assets of the underlying ERC20 token being withdrawn.
+     * @param receiver The address of the receiver of the underlying ERC20 tokens.
+     * @param owner_ The address of the owner of the assets being withdrawn.
+     * @return shares The corresponding amount of shares redeemed.
+     */
+    function withdraw(uint256 assets, address receiver, address owner_) public override returns (uint256 shares) {
+        shares = ERC4626(TRANCHE).withdraw(assets, address(this), address(this));
+
+        if (msg.sender != owner_) {
+            // Saves gas for limited approvals.
+            uint256 allowed = allowance[owner_][msg.sender];
+
+            if (allowed != type(uint256).max) allowance[owner_][msg.sender] = allowed - shares;
+        }
+
+        _burn(owner_, shares);
+
+        asset.safeTransfer(receiver, assets);
+
+        emit Withdraw(msg.sender, receiver, owner_, assets, shares);
+    }
+
+    /**
+     * @notice Compliant with the standard ERC-4626 redeem implementation.
+     * @param shares The amount of shares being redeemed.
+     * @param receiver The address of the receiver of the underlying ERC20 tokens.
+     * @param owner_ The address of the owner of the shares being redeemed.
+     * @return assets The corresponding amount of assets withdrawn.
+     */
+    function redeem(uint256 shares, address receiver, address owner_) public override returns (uint256 assets) {
+        assets = ERC4626(TRANCHE).redeem(shares, address(this), address(this));
+
+        if (msg.sender != owner_) {
```

Comment:
Allowance check should be done before calling `redeem` on the tranche. As here we already have the `shares` needed for the check.